### DISCUSSION TITLE
Have a way to get failure from publish method if mqtt broker not reachable

### DISCUSSION THREAD

**Discussion by martinscheffler** [empicano/aiomqtt#77] [Ideas]
Hi, 
I have a scenario where no message may get lost. It would be good if there was a way to call publish and have it fail if it cannot be delivered immediately.

**Comment by frederikaalund** [comment]
Hi Martin

Thanks for opening this discussion.

> I have a scenario where no message may get lost.

Sounds like a good fit for [QoS level 1 and above](https://www.hivemq.com/blog/mqtt-essentials-part-6-mqtt-quality-of-service-levels/).

> It would be good if there was a way to call publish and have it fail if it cannot be delivered immediately.

When `await client.publish(topic, message, qos=1)` returns, the message is at the broker (all send/receive/handshaking is done). That's a guarantee from asyncio-mqtt (for QoS level 1 and above). :+1: 

From there on, it's the broker's responsibility to send the message to the clients. There is nothing asyncio-mqtt can guarantee about that. It's up to the broker.

If asyncio-mqtt can't publish the message, then `await client.publish(topic, message, qos=1)` raises an exception. The default timeout is 10 seconds and you can change it with the `timeout` keyword argument.

Does this fit your use case?

~Frederik
